homeassistant/components/moscow_yandex_transport/sensor.py

```diff
@@ -0,0 +1,178 @@
+    def __init__(self, requester, stop_id, routes, name):
+        """
+
+        :type requester: data provider for request to yandex api
+        """
+        self.requester = requester
+        self._stop_id = stop_id
+        self._routes = []
+        for route in routes:
+            self._routes.append(str(route))
+        self._state = None
+        self._name = name
+        self._attrs = None
+        self._next_route = None
+
+    def update(self):
+        """Get the latest data from maps.yandex.ru and update the states."""
+        result = {}
+        closer_time = None
+        try:
+            yandex_reply = self.requester.get_stop_info(self._stop_id)
+            data = yandex_reply["data"]
+            stop_metadata = data["properties"]["StopMetaData"]
+        except KeyError as e:
+            _LOGGER.warning(f"Exception KeyError was captured, missing key is {e}. Yandex returned :{yandex_reply}")
+            self.requester.set_new_session()
+            data = self.requester.get_stop_info(self._stop_id)["data"]
+            stop_metadata = data["properties"]["StopMetaData"]
+        stop_name = data["properties"]["name"]
+        transport_list = stop_metadata["Transport"]
+        for transport in transport_list:
+            route = transport["name"]
+            if self._routes and route not in self._routes:
+                # skip unnecessary route info
+                continue
+            if "Events" in transport["BriefSchedule"]:
+                for event in transport["BriefSchedule"]["Events"]:
+                    if "Estimated" in event:
+                        posix_time_next = int(event["Estimated"]["value"])
+                        if closer_time is None or closer_time > posix_time_next:
+                            closer_time = posix_time_next
+                        if route not in result:
+                            result[route] = []
+                        result[route].append(event["Estimated"]["text"])
+        for route in result:
+            result[route] = ", ".join(result[route])
+        result[STOP_NAME] = stop_name
+        result[ATTR_ATTRIBUTION] = ATTRIBUTION
+        if closer_time is None:
+            self._state = "n/a"
+        else:
+            self._state = due_in_minutes(closer_time)
+        self._attrs = result
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attrs
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit this state is expressed in."""
+        return "min"
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return ICON
+
+
+class YandexMapsRequester(object):
```

Comment:
Please extract the service specific interface code, I/O and data parsing, to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services

Reply:
Done